File: source-connector/src/main/java/sourceconnector/repository/file/S3FileLister.java
package sourceconnector.repository.file;

import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
import software.amazon.awssdk.services.s3.model.S3Object;
import sourceconnector.domain.file.FileKey;
import sourceconnector.repository.file.validator.FileValidator;

import javax.swing.*;
import java.util.ArrayList;
import java.util.List;

public class S3FileLister implements FileLister  {
  private final FileValidator fileValidator;
  private final S3Client s3Client;

  // FIXME: No need bucket name
  public S3FileLister(
    S3Client s3Client,
    FileValidator fileValidator
  ) {
    this.s3Client = s3Client;
    this.fileValidator = fileValidator;
  }

  // TODO: Refactor for using S3 URI creator
  /**
   * Get all s3 object key paths <br>
   * this can handle both directory and file path
   * @return {@code List<String>}
   */
  @Override
  public List<FileKey> listFiles(FileKey... inputFileKeys) {

    List<FileKey> fileKeys = new ArrayList<>();
    for (FileKey fileKey : inputFileKeys) {
      S3Location s3Location = S3Location.from(fileKey);
       ListObjectsV2Request request = ListObjectsV2Request.builder()
        .bucket(s3Location.bucket())
        .prefix(s3Location.key())
        .delimiter("/")
        .build();

      List<FileKey> keys = this.s3Client.listObjectsV2Paginator(request)
        .stream()
        .flatMap(response -> response.contents().stream())
        .map(s3Object -> new S3Location(s3Location.bucket(), s3Object.key()).toFileKey())
        .filter(fileValidator::isValid)
        .toList();

      fileKeys.addAll(keys);
    }

    return fileKeys;
  }

  @Override
  public List<FileKey> listFilesRecursively(FileKey... inputFileKeys) {

    List<FileKey> fileKeys = new ArrayList<>();
    for (FileKey fileKey : inputFileKeys) {
      S3Location s3Location = S3Location.from(fileKey);
      ListObjectsV2Request request = ListObjectsV2Request.builder()
        .bucket(s3Location.bucket())
        .prefix(s3Location.key())
        .build();

      List<FileKey> keys = this.s3Client.listObjectsV2Paginator(request)
        .stream()
        .flatMap(response -> response.contents().stream())
        .map(s3Object -> new S3Location(s3Location.bucket(), s3Object.key()).toFileKey())
        .filter(fileValidator::isValid)
        .toList();

      fileKeys.addAll(keys);
    }

    return fileKeys;
  }
}

File: source-connector/src/main/kotlin/sourceconnector/repository/file/S3FileLister.kt
package sourceconnector.repository.file

import offsetmanager.domain.file.FileKey
import software.amazon.awssdk.services.s3.S3Client
import software.amazon.awssdk.services.s3.model.ListObjectsV2Request
import software.amazon.awssdk.services.s3.model.ListObjectsV2Response
import sourceconnector.repository.file.S3Location.Companion.from
import sourceconnector.repository.file.validator.FileValidator


class S3FileLister(
  private val s3Client: S3Client,
  private val fileValidator: FileValidator
) : FileLister {
  /**
   * Get all s3 object key paths <br></br>
   * this can handle both directory and file path
   * @return `List<String>`
   */
  override fun listFiles(vararg inputFileKeys: FileKey): List<FileKey> {
    val fileKeys: MutableList<FileKey> = mutableListOf()
    for (fileKey in inputFileKeys) {
      val s3Location = from(fileKey)

      val request = ListObjectsV2Request.builder()
        .bucket(s3Location.bucket)
        .prefix(s3Location.key)
        .delimiter("/")
        .build()

      val keys = this.listFilesInResponse(fileKey, request)

      fileKeys.addAll(keys)
    }

    return fileKeys
  }

  override fun listFilesRecursively(vararg inputFileKeys: FileKey): List<FileKey> {
    val fileKeys: MutableList<FileKey> = mutableListOf()
    for (fileKey in inputFileKeys) {
      val s3Location = from(fileKey)
      val request = ListObjectsV2Request.builder()
        .bucket(s3Location.bucket)
        .prefix(s3Location.key)
        .build()

      val keys = this.listFilesInResponse(fileKey, request)

      fileKeys.addAll(keys)
    }

    return fileKeys
  }

  private fun listFilesInResponse(fileKey: FileKey, request: ListObjectsV2Request): List<FileKey> {
    val s3Location = from(fileKey)

    return this.s3Client.listObjectsV2Paginator(request)
      .flatMap { response: ListObjectsV2Response? -> response!!.contents() }
      .map { s3Object -> S3Location(s3Location.bucket, s3Object.key()).toFileKey() }
      .filter { fileKey -> fileValidator.isValid(fileKey) }
      .toList()
  }
}
